File: Python/binarysearchtree.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    # Insert a new key
    def insert(self, key):
        self.root = self._insert(self.root, key)

    def _insert(self, root, key):
        if root is None:
            return Node(key)
        if key < root.key:
            root.left = self._insert(root.left, key)
        elif key > root.key:
            root.right = self._insert(root.right, key)
        return root

    # Search for a key
    def search(self, key):
        return self._search(self.root, key)

    def _search(self, root, key):
        if root is None or root.key == key:
            return root
        if key < root.key:
            return self._search(root.left, key)
        return self._search(root.right, key)

    # Delete a key
    def delete(self, key):
        self.root = self._delete(self.root, key)

    def _delete(self, root, key):
        if root is None:
            return root

        if key < root.key:
            root.left = self._delete(root.left, key)
        elif key > root.key:
            root.right = self._delete(root.right, key)
        else:
            # Node with only one child or no child
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            # Node with two children
            min_node = self._minValueNode(root.right)
            root.key = min_node.key
            root.right = self._delete(root.right, min_node.key)
        return root

    def _minValueNode(self, node):
        current = node
        while current.left:
            current = current.left
        return current

    # Inorder Traversal (sorted order)
    def inorder(self):
        return self._inorder(self.root)

    def _inorder(self, root):
        return self._inorder(root.left) + [root.key] + self._inorder(root.right) if root else []


# Example usage
if __name__ == "__main__":
    bst = BST()
    elements = list(map(int, input("Enter numbers to insert (space-separated): ").split()))
    for el in elements:
        bst.insert(el)

    print("Inorder Traversal:", bst.inorder())

    key = int(input("Enter a value to search: "))
    print("Found!" if bst.search(key) else "Not found.")

    delete_key = int(input("Enter a value to delete: "))
    bst.delete(delete_key)
    print("Inorder Traversal after deletion:", bst.inorder())

File: javascript/binary_search_tree.js
/**
 * Binary Search Tree Implementation in JavaScript
 * ===============================================
 * 
 * Complete BST with insertion, deletion, traversal, and utility operations.
 * Includes balancing check, height calculation, and path finding.
 * 
 * @author Hacktoberfest 2025 Contributor
 */

class TreeNode {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }

    /**
     * Insert a value into the BST
     * Time: O(log n) average, O(n) worst
     * Space: O(1)
     */
    insert(value) {
        const newNode = new TreeNode(value);
        
        if (!this.root) {
            this.root = newNode;
            return this;
        }
        
        let current = this.root;
        while (true) {
            if (value === current.value) return undefined; // Duplicate
            
            if (value < current.value) {
                if (!current.left) {
                    current.left = newNode;
                    return this;
                }
                current = current.left;
            } else {
                if (!current.right) {
                    current.right = newNode;
                    return this;
                }
                current = current.right;
            }
        }
    }

    /**
     * Search for a value
     * Time: O(log n) average
     */
    search(value) {
        let current = this.root;
        
        while (current) {
            if (value === current.value) return true;
            if (value < current.value) current = current.left;
            else current = current.right;
        }
        
        return false;
    }

    /**
     * Find minimum value
     */
    findMin(node = this.root) {
        if (!node) return null;
        while (node.left) node = node.left;
        return node.value;
    }

    /**
     * Find maximum value
     */
    findMax(node = this.root) {
        if (!node) return null;
        while (node.right) node = node.right;
        return node.value;
    }

    /**
     * Delete a value from BST
     * Time: O(log n) average
     */
    delete(value) {
        this.root = this._deleteNode(this.root, value);
        return this;
    }

    _deleteNode(node, value) {
        if (!node) return null;
        
        if (value < node.value) {
            node.left = this._deleteNode(node.left, value);
        } else if (value > node.value) {
            node.right = this._deleteNode(node.right, value);
        } else {
            // Node to delete found
            
            // Case 1: Leaf node
            if (!node.left && !node.right) return null;
            
            // Case 2: One child
            if (!node.left) return node.right;
            if (!node.right) return node.left;
            
            // Case 3: Two children
            const minRight = this.findMin(node.right);
            node.value = minRight;
            node.right = this._deleteNode(node.right, minRight);
        }
        
        return node;
    }

    /**
     * In-order traversal (Left-Root-Right)
     * Returns sorted array
     */
    inOrder(node = this.root, result = []) {
        if (node) {
            this.inOrder(node.left, result);
            result.push(node.value);
            this.inOrder(node.right, result);
        }
        return result;
    }

    /**
     * Pre-order traversal (Root-Left-Right)
     */
    preOrder(node = this.root, result = []) {
        if (node) {
            result.push(node.value);
            this.preOrder(node.left, result);
            this.preOrder(node.right, result);
        }
        return result;
    }

    /**
     * Post-order traversal (Left-Right-Root)
     */
    postOrder(node = this.root, result = []) {
        if (node) {
            this.postOrder(node.left, result);
            this.postOrder(node.right, result);
            result.push(node.value);
        }
        return result;
    }

    /**
     * Level-order traversal (BFS)
     */
    levelOrder() {
        if (!this.root) return [];
        
        const result = [];
        const queue = [this.root];
        
        while (queue.length) {
            const node = queue.shift();
            result.push(node.value);
            
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        
        return result;
    }

    /**
     * Calculate height of tree
     * Time: O(n)
     */
    height(node = this.root) {
        if (!node) return -1;
        return 1 + Math.max(this.height(node.left), this.height(node.right));
    }

    /**
     * Check if tree is balanced
     */
    isBalanced(node = this.root) {
        if (!node) return true;
        
        const leftHeight = this.height(node.left);
        const rightHeight = this.height(node.right);
        
        return Math.abs(leftHeight - rightHeight) <= 1 &&
               this.isBalanced(node.left) &&
               this.isBalanced(node.right);
    }

    /**
     * Count total nodes
     */
    size(node = this.root) {
        if (!node) return 0;
        return 1 + this.size(node.left) + this.size(node.right);
    }

    /**
     * Find path to a value
     */
    findPath(value, node = this.root, path = []) {
        if (!node) return null;
        
        path.push(node.value);
        
        if (node.value === value) return path;
        
        if (value < node.value) {
            return this.findPath(value, node.left, path);
        } else {
            return this.findPath(value, node.right, path);
        }
    }

    /**
     * Validate if tree is a valid BST
     */
    isValidBST(node = this.root, min = -Infinity, max = Infinity) {
        if (!node) return true;
        
        if (node.value <= min || node.value >= max) return false;
        
        return this.isValidBST(node.left, min, node.value) &&
               this.isValidBST(node.right, node.value, max);
    }
}

// Demonstration
console.log('='.repeat(70));
console.log('BINARY SEARCH TREE DEMONSTRATION');
console.log('='.repeat(70));

const bst = new BinarySearchTree();

// Insert values
console.log('\n1. INSERTION');
[50, 30, 70, 20, 40, 60, 80].forEach(val => bst.insert(val));
console.log('Inserted: 50, 30, 70, 20, 40, 60, 80');

// Traversals
console.log('\n2. TRAVERSALS');
console.log('In-order:', bst.inOrder());
console.log('Pre-order:', bst.preOrder());
console.log('Post-order:', bst.postOrder());
console.log('Level-order:', bst.levelOrder());

// Search
console.log('\n3. SEARCH');
console.log('Search 40:', bst.search(40));
console.log('Search 100:', bst.search(100));

// Properties
console.log('\n4. TREE PROPERTIES');
console.log('Height:', bst.height());
console.log('Size:', bst.size());
console.log('Min:', bst.findMin());
console.log('Max:', bst.findMax());
console.log('Is Balanced:', bst.isBalanced());
console.log('Is Valid BST:', bst.isValidBST());

// Path finding
console.log('\n5. PATH FINDING');
console.log('Path to 40:', bst.findPath(40));

// Deletion
console.log('\n6. DELETION');
bst.delete(30);
console.log('After deleting 30:', bst.inOrder());

console.log('\n' + '='.repeat(70));

module.exports = { TreeNode, BinarySearchTree };
